File: internal/level1_tests/pcie_width_missing_lanes_check.go
package level1_tests

import (
	"fmt"
	"github.com/oracle/oci-dr-hpc-v2/internal/test_limits"
	"regexp"
	"strconv"
	"strings"

	"github.com/oracle/oci-dr-hpc-v2/internal/executor"
	"github.com/oracle/oci-dr-hpc-v2/internal/logger"
	"github.com/oracle/oci-dr-hpc-v2/internal/reporter"
)

// PCIeWidthResult represents PCIe width analysis results
type PCIeWidthResult struct {
	WidthCounts map[string]int `json:"width_counts"`
	SpeedCounts map[string]int `json:"speed_counts"`
	StateErrors []string       `json:"state_errors"`
	Success     bool           `json:"success"`
	ErrorMsg    string         `json:"error_message,omitempty"`
}

// PCIeWidthMissingLanesTestConfig holds configuration for this test
type PCIeWidthMissingLanesTestConfig struct {
	IsEnabled            bool               `json:"enabled"`
	ExpectedGPUWidths    map[string]int     `json:"expected_gpu_widths"`
	ExpectedRDMAWidths   map[string]int     `json:"expected_rdma_widths"`
	ExpectedGPUSpeeds    map[string]int     `json:"expected_gpu_speeds"`
	ExpectedRDMASpeeds   map[string]int     `json:"expected_rdma_speeds"`
	ExpectedLinkState    string             `json:"expected_link_state"`
}

// getpcieWidthMissingLanesTestConfig loads test configuration
func getPcieWidthMissingLanesTestConfig(shape string) (*PCIeWidthMissingLanesTestConfig, error) {
	// Load configuration from test_limits.json
	limits, err := test_limits.LoadTestLimits()
	if err != nil {
		return nil, err
	}

	// Initialize config with defaults
	config := &PCIeWidthMissingLanesTestConfig{
		IsEnabled:           false,
		ExpectedGPUWidths:   make(map[string]int),
		ExpectedRDMAWidths:  make(map[string]int),
		ExpectedGPUSpeeds:   make(map[string]int),
		ExpectedRDMASpeeds:  make(map[string]int),
		ExpectedLinkState:   "Ok",
	}

	// Check if test is enabled
	enabled, err := limits.IsTestEnabled(shape, "pcie_width_missing_lanes_check")
	if err != nil {
		return nil, err
	}
	config.IsEnabled = enabled

	// Get expected GPU widths
	gpuWidthsData, err := limits.GetThresholdForTest(shape, "pcie_width_missing_lanes_check")
	if err != nil {
		return nil, err
	}

	// Parse the threshold data
	if thresholdMap, ok := gpuWidthsData.(map[string]interface{}); ok {
		// Parse GPU widths
		if gpuWidths, exists := thresholdMap["gpu_widths"]; exists {
			if gpuMap, ok := gpuWidths.(map[string]interface{}); ok {
				for width, count := range gpuMap {
					if countFloat, ok := count.(float64); ok {
						config.ExpectedGPUWidths[width] = int(countFloat)
					}
				}
			}
		}

		// Parse RDMA widths
		if rdmaWidths, exists := thresholdMap["rdma_widths"]; exists {
			if rdmaMap, ok := rdmaWidths.(map[string]interface{}); ok {
				for width, count := range rdmaMap {
					if countFloat, ok := count.(float64); ok {
						config.ExpectedRDMAWidths[width] = int(countFloat)
					}
				}
			}
		}

		// Parse GPU speeds
		if gpuSpeeds, exists := thresholdMap["gpu_speeds"]; exists {
			if gpuSpeedMap, ok := gpuSpeeds.(map[string]interface{}); ok {
				for speed, count := range gpuSpeedMap {
					if countFloat, ok := count.(float64); ok {
						config.ExpectedGPUSpeeds[speed] = int(countFloat)
					}
				}
			}
		}

		// Parse RDMA speeds
		if rdmaSpeeds, exists := thresholdMap["rdma_speeds"]; exists {
			if rdmaSpeedMap, ok := rdmaSpeeds.(map[string]interface{}); ok {
				for speed, count := range rdmaSpeedMap {
					if countFloat, ok := count.(float64); ok {
						config.ExpectedRDMASpeeds[speed] = int(countFloat)
					}
				}
			}
		}

		// Parse expected link state
		if linkState, exists := thresholdMap["expected_link_state"]; exists {
			if stateStr, ok := linkState.(string); ok {
				config.ExpectedLinkState = stateStr
			}
		}
	}

	return config, nil
}

// PCIeParseResult holds parsed PCIe information
type PCIeParseResult struct {
	WidthCounts map[string]int
	SpeedCounts map[string]int
	StateErrors []string
}

// FilterLspciForDevice filters lspci output to extract LnkSta lines for specific device type
func FilterLspciForDevice(lspciOutput string, deviceType string) string {
	lines := strings.Split(lspciOutput, "\n")
	var filteredLines []string
	var deviceFound bool
	
	devicePattern := strings.ToLower(deviceType)
	
	for _, line := range lines {
		// Check if this line is a PCI device header (starts with bus:device.function)
		if regexp.MustCompile(`^[0-9a-fA-F]{2}:[0-9a-fA-F]{2}\.[0-9a-fA-F]`).MatchString(line) {
			// Reset device found flag
			deviceFound = false
			
			// Check if this device matches our target type
			lowerLine := strings.ToLower(line)
			switch devicePattern {
			case "nvidia", "gpu", "nvswitch":
				deviceFound = strings.Contains(lowerLine, "nvidia")
			case "mellanox", "rdma":
				deviceFound = strings.Contains(lowerLine, "mellanox")
			}
		}
		
		// If we found a matching device and this line contains LnkSta, include it
		if deviceFound && strings.Contains(line, "LnkSta:") {
			filteredLines = append(filteredLines, line)
		}
	}
	
	return strings.Join(filteredLines, "\n")
}

// AggregateLinkStatistics aggregates LnkSta lines by counting identical entries
func AggregateLinkStatistics(linkStatusLines string) string {
	if strings.TrimSpace(linkStatusLines) == "" {
		return ""
	}
	
	lines := strings.Split(strings.TrimSpace(linkStatusLines), "\n")
	countMap := make(map[string]int)
	
	// Count occurrences of each unique LnkSta line
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line != "" {
			countMap[line]++
		}
	}
	
	// Convert back to the format expected by parseLspciWidthOutput
	var result []string
	for line, count := range countMap {
		result = append(result, fmt.Sprintf("%d\t%s", count, line))
	}
	
	return strings.Join(result, "\n")
}

// parseLspciWidthOutput parses lspci output to extract PCIe width, speed, and state information
func parseLspciWidthOutput(output string, expectedLinkState string) PCIeParseResult {
	result := PCIeParseResult{
		WidthCounts: make(map[string]int),
		SpeedCounts: make(map[string]int),
		StateErrors: []string{},
	}
	
	lines := strings.Split(strings.TrimSpace(output), "\n")
	
	// Enhanced regex to capture count, speed with (ok), width with (ok)
	// Example: "4       LnkSta: Speed 16GT/s (ok), Width x16 (ok)"
	re := regexp.MustCompile(`^\s*(\d+)\s+LnkSta:\s*Speed\s+([^\s]+)\s*\(([^)]+)\),\s*Width\s+x(\d+)\s*\(([^)]+)\)`)
	
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		
		matches := re.FindStringSubmatch(line)
		if len(matches) == 6 {
			count, err := strconv.Atoi(matches[1])
			if err != nil {
				logger.Debugf("Failed to parse count from line: %s", line)
				continue
			}
			
			speed := strings.TrimSpace(matches[2])
			speedState := strings.TrimSpace(matches[3])
			width := matches[4]
			widthState := strings.TrimSpace(matches[5])
			
			// Count widths only if width state is ok
			widthKey := fmt.Sprintf("Width x%s", width)
			if widthState == "ok" {
				result.WidthCounts[widthKey] += count
			} else {
				result.StateErrors = append(result.StateErrors, 
					fmt.Sprintf("%d devices have width state '%s' instead of 'ok'", count, widthState))
			}
			
			// Count speeds only if speed state is ok
			speedKey := fmt.Sprintf("Speed %s", speed)
			if speedState == "ok" {
				result.SpeedCounts[speedKey] += count
			} else {
				result.StateErrors = append(result.StateErrors, 
					fmt.Sprintf("%d devices have speed state '%s' instead of 'ok'", count, speedState))
			}
			
			logger.Debugf("Parsed PCIe: %s (%s) = %d, %s (%s) = %d", widthKey, widthState, count, speedKey, speedState, count)
		}
	}
	
	return result
}

// checkGPUNVSwitchPCIeWidth checks PCIe width, speed, and state for GPU and NVSwitch interfaces
func checkGPUNVSwitchPCIeWidth(expectedLinkState string) PCIeWidthResult {
	logger.Info("Checking GPU/NVSwitch PCIe width, speed, and state...")
	
	// Execute the lspci command with verbose output
	result, err := executor.RunLspci("-vvv")
	if err != nil {
		return PCIeWidthResult{
			Success:  false,
			ErrorMsg: fmt.Sprintf("Failed to execute lspci command: %v", err),
		}
	}
	
	if result.ExitCode != 0 {
		return PCIeWidthResult{
			Success:  false,
			ErrorMsg: fmt.Sprintf("lspci command failed with exit code %d: %s", result.ExitCode, result.Output),
		}
	}
	
	// Filter lspci output for NVIDIA devices and extract LnkSta lines
	filteredOutput := FilterLspciForDevice(result.Output, "nvidia")
	if strings.TrimSpace(filteredOutput) == "" {
		return PCIeWidthResult{
			Success:  false,
			ErrorMsg: "No NVIDIA PCIe devices found",
		}
	}
	
	// Aggregate identical LnkSta lines with counts
	aggregatedOutput := AggregateLinkStatistics(filteredOutput)
	
	parseResult := parseLspciWidthOutput(aggregatedOutput, expectedLinkState)
	
	return PCIeWidthResult{
		WidthCounts: parseResult.WidthCounts,
		SpeedCounts: parseResult.SpeedCounts,
		StateErrors: parseResult.StateErrors,
		Success:     true,
	}
}

// checkRDMAPCIeWidth checks PCIe width, speed, and state for RDMA interfaces
func checkRDMAPCIeWidth(expectedLinkState string) PCIeWidthResult {
	logger.Info("Checking RDMA PCIe width, speed, and state...")
	
	// Execute the lspci command with verbose output
	result, err := executor.RunLspci("-vvv")
	if err != nil {
		return PCIeWidthResult{
			Success:  false,
			ErrorMsg: fmt.Sprintf("Failed to execute lspci command: %v", err),
		}
	}
	
	if result.ExitCode != 0 {
		return PCIeWidthResult{
			Success:  false,
			ErrorMsg: fmt.Sprintf("lspci command failed with exit code %d: %s", result.ExitCode, result.Output),
		}
	}
	
	// Filter lspci output for Mellanox devices and extract LnkSta lines
	filteredOutput := FilterLspciForDevice(result.Output, "mellanox")
	if strings.TrimSpace(filteredOutput) == "" {
		return PCIeWidthResult{
			Success:  false,
			ErrorMsg: "No Mellanox PCIe devices found",
		}
	}
	
	// Aggregate identical LnkSta lines with counts
	aggregatedOutput := AggregateLinkStatistics(filteredOutput)
	
	parseResult := parseLspciWidthOutput(aggregatedOutput, expectedLinkState)
	
	return PCIeWidthResult{
		WidthCounts: parseResult.WidthCounts,
		SpeedCounts: parseResult.SpeedCounts,
		StateErrors: parseResult.StateErrors,
		Success:     true,
	}
}

// validateWidthCounts compares actual vs expected width counts
func validateWidthCounts(actual, expected map[string]int, deviceType string) (bool, string) {
	for width, expectedCount := range expected {
		actualCount, exists := actual[width]
		if !exists {
			actualCount = 0
		}
		
		if actualCount != expectedCount {
			return false, fmt.Sprintf("%s PCIe width mismatch: expected %dx %s, got %dx %s", 
				deviceType, expectedCount, width, actualCount, width)
		}
	}
	
	return true, ""
}

// validateSpeedCounts compares actual vs expected speed counts
func validateSpeedCounts(actual, expected map[string]int, deviceType string) (bool, string) {
	for speed, expectedCount := range expected {
		actualCount, exists := actual[speed]
		if !exists {
			actualCount = 0
		}
		
		if actualCount != expectedCount {
			return false, fmt.Sprintf("%s PCIe speed mismatch: expected %dx %s, got %dx %s", 
				deviceType, expectedCount, speed, actualCount, speed)
		}
	}
	
	return true, ""
}

// RunPCIeWidthMissingLanesCheck performs the PCIe width missing lanes diagnostic
func RunPCIeWidthMissingLanesCheck() error {
	logger.Info("=== PCIe Width Missing Lanes Check ===")
	rep := reporter.GetReporter()

	// Step 1: Get shape from IMDS
	logger.Info("Step 1: Getting shape from IMDS...")
	shape, err := executor.GetCurrentShape()
	if err != nil {
		logger.Error("PCIe Width Missing Lanes Check: FAIL - Could not get shape from IMDS:", err)
		rep.AddPCIeWidthResult("FAIL", nil, nil, nil, nil, nil, err)
		return fmt.Errorf("failed to get shape from IMDS: %w", err)
	}
	logger.Info("Current shape from IMDS:", shape)

	// Step 2: Load test configuration
	config, err := getPcieWidthMissingLanesTestConfig(shape)
	if err != nil {
		logger.Error("PCIe Width Missing Lanes Check: FAIL - Could not load test configuration:", err)
		rep.AddPCIeWidthResult("FAIL", nil, nil, nil, nil, nil, err)
		return fmt.Errorf("failed to load test configuration: %w", err)
	}

	if !config.IsEnabled {
		errorMsg := fmt.Sprintf("Test not applicable for this shape %s", shape)
		logger.Error(errorMsg)
		rep.AddPCIeWidthResult("SKIP", nil, nil, nil, nil, nil, fmt.Errorf(errorMsg))
		return fmt.Errorf(errorMsg)
	}

	// Step 3: Check GPU/NVSwitch PCIe width, speed, and state
	logger.Info("Step 2: Checking GPU/NVSwitch PCIe width, speed, and state...")
	gpuResult := checkGPUNVSwitchPCIeWidth(config.ExpectedLinkState)
	
	var errorMessages []string
	var allGPUWidths map[string]int
	var allRDMAWidths map[string]int
	var allGPUSpeeds map[string]int
	var allRDMASpeeds map[string]int
	var allStateErrors []string

	if !gpuResult.Success {
		errorMessages = append(errorMessages, fmt.Sprintf("GPU/NVSwitch: %s", gpuResult.ErrorMsg))
		allGPUWidths = make(map[string]int)
		allGPUSpeeds = make(map[string]int)
	} else {
		allGPUWidths = gpuResult.WidthCounts
		allGPUSpeeds = gpuResult.SpeedCounts
		allStateErrors = append(allStateErrors, gpuResult.StateErrors...)
		
		// Validate GPU widths
		if len(config.ExpectedGPUWidths) > 0 {
			gpuValid, gpuError := validateWidthCounts(allGPUWidths, config.ExpectedGPUWidths, "GPU/NVSwitch")
			if !gpuValid {
				errorMessages = append(errorMessages, gpuError)
			}
		}
		
		// Validate GPU speeds
		if len(config.ExpectedGPUSpeeds) > 0 {
			gpuSpeedValid, gpuSpeedError := validateSpeedCounts(gpuResult.SpeedCounts, config.ExpectedGPUSpeeds, "GPU/NVSwitch")
			if !gpuSpeedValid {
				errorMessages = append(errorMessages, gpuSpeedError)
			}
		}
		
		// Check GPU state errors
		if len(gpuResult.StateErrors) > 0 {
			for _, stateError := range gpuResult.StateErrors {
				errorMessages = append(errorMessages, fmt.Sprintf("GPU/NVSwitch state error: %s", stateError))
			}
		}
	}

	// Step 4: Check RDMA PCIe width, speed, and state
	logger.Info("Step 3: Checking RDMA PCIe width, speed, and state...")
	rdmaResult := checkRDMAPCIeWidth(config.ExpectedLinkState)
	
	if !rdmaResult.Success {
		errorMessages = append(errorMessages, fmt.Sprintf("RDMA: %s", rdmaResult.ErrorMsg))
		allRDMAWidths = make(map[string]int)
		allRDMASpeeds = make(map[string]int)
	} else {
		allRDMAWidths = rdmaResult.WidthCounts
		allRDMASpeeds = rdmaResult.SpeedCounts
		allStateErrors = append(allStateErrors, rdmaResult.StateErrors...)
		
		// Validate RDMA widths
		if len(config.ExpectedRDMAWidths) > 0 {
			rdmaValid, rdmaError := validateWidthCounts(allRDMAWidths, config.ExpectedRDMAWidths, "RDMA")
			if !rdmaValid {
				errorMessages = append(errorMessages, rdmaError)
			}
		}
		
		// Validate RDMA speeds
		if len(config.ExpectedRDMASpeeds) > 0 {
			rdmaSpeedValid, rdmaSpeedError := validateSpeedCounts(rdmaResult.SpeedCounts, config.ExpectedRDMASpeeds, "RDMA")
			if !rdmaSpeedValid {
				errorMessages = append(errorMessages, rdmaSpeedError)
			}
		}
		
		// Check RDMA state errors
		if len(rdmaResult.StateErrors) > 0 {
			for _, stateError := range rdmaResult.StateErrors {
				errorMessages = append(errorMessages, fmt.Sprintf("RDMA state error: %s", stateError))
			}
		}
	}

	// Step 5: Generate final result
	if len(errorMessages) == 0 {
		logger.Info("PCIe Width Missing Lanes Check: PASS - All PCIe interfaces operating at expected width and speed")
		rep.AddPCIeWidthResult("PASS", allGPUWidths, allRDMAWidths, allGPUSpeeds, allRDMASpeeds, allStateErrors, nil)
		return nil
	} else {
		// Combine error messages
		combinedError := strings.Join(errorMessages, "; ") + ". Please reboot the host and if the issue persists, send the node to OCI."
		logger.Error("PCIe Width Missing Lanes Check: FAIL -", combinedError)
		
		err := fmt.Errorf(combinedError)
		rep.AddPCIeWidthResult("FAIL", allGPUWidths, allRDMAWidths, allGPUSpeeds, allRDMASpeeds, allStateErrors, err)
		return err
	}
}

File: scripts/BM.GPU.H100.8/pcie_width_missing_lanes_check.py
#!/usr/bin/env python3

"""
PCIe Width Missing Lanes Check

This script checks the PCIe link width for GPU and RDMA interfaces to detect missing lanes.
It verifies that all PCIe links are operating at their expected width configuration.

Expected output for BM.GPU.H100.8:
- GPU/NVSwitch: 4x Width x2 (ok), 8x Width x16 (ok)  
- RDMA: 2x Width x8 (ok), 16x Width x16 (ok)

Author: Oracle Cloud Infrastructure
"""

import json
import os
import subprocess
import sys
from datetime import datetime
from typing import Dict, List, Tuple


def run_command(cmd: str) -> Tuple[int, str, str]:
    """
    Execute a shell command and return exit code, stdout, stderr.
    
    Args:
        cmd: Command to execute
        
    Returns:
        Tuple of (exit_code, stdout, stderr)
    """
    try:
        result = subprocess.run(
            cmd, 
            shell=True, 
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE, 
            universal_newlines=True, 
            timeout=30
        )
        return result.returncode, result.stdout.strip(), result.stderr.strip()
    except subprocess.TimeoutExpired:
        return 124, "", "Command timed out"
    except Exception as e:
        return 1, "", str(e)


def parse_lspci_width_output(output: str) -> Tuple[Dict[str, int], Dict[str, int], List[str]]:
    """
    Parse lspci output to extract PCIe width, speed, and state information.
    
    Args:
        output: Raw output from lspci command
        
    Returns:
        Tuple of (width_counts, speed_counts, state_errors)
    """
    width_counts = {}
    speed_counts = {}
    state_errors = []
    
    lines = output.strip().split('\n')
    for line in lines:
        line = line.strip()
        if not line:
            continue
            
        # Parse line format: "count           LnkSta: Speed 16GT/s (ok), Width x16 (ok)"
        import re
        match = re.match(r'^\s*(\d+)\s+LnkSta:\s*Speed\s+([^\s]+)\s*\(([^)]+)\),\s*Width\s+x(\d+)\s*\(([^)]+)\)', line)
        if match:
            try:
                count = int(match.group(1))
                speed = match.group(2)
                speed_state = match.group(3)
                width = match.group(4)
                width_state = match.group(5)
                
                # Count widths only if width state is ok
                width_key = f"Width x{width}"
                if width_state == "ok":
                    width_counts[width_key] = width_counts.get(width_key, 0) + count
                else:
                    state_errors.append(f"{count} devices have width state '{width_state}' instead of 'ok'")
                
                # Count speeds only if speed state is ok
                speed_key = f"Speed {speed}"
                if speed_state == "ok":
                    speed_counts[speed_key] = speed_counts.get(speed_key, 0) + count
                else:
                    state_errors.append(f"{count} devices have speed state '{speed_state}' instead of 'ok'")
                    
            except (ValueError, IndexError):
                continue
                
    return width_counts, speed_counts, state_errors


def check_gpu_nvswitch_pcie_width() -> Tuple[bool, str, Dict[str, int], Dict[str, int], List[str]]:
    """
    Check PCIe width, speed, and state for GPU and NVSwitch interfaces.
    
    Returns:
        Tuple of (success, error_message, width_counts, speed_counts, state_errors)
    """
    cmd = "sudo lspci -vvv | egrep '^[0-9,a-f]|LnkSta:' | grep -A 1 -i nvidia | grep LnkSta | sort | uniq -c"
    
    exit_code, stdout, stderr = run_command(cmd)
    
    if exit_code != 0:
        return False, f"Failed to execute lspci command: {stderr}", {}, {}, []
    
    if not stdout.strip():
        return False, "No NVIDIA PCIe devices found", {}, {}, []
    
    width_counts, speed_counts, state_errors = parse_lspci_width_output(stdout)
    
    # Expected for BM.GPU.H100.8: 4x Width x2, 8x Width x16
    expected_width_counts = {
        "Width x2": 4,
        "Width x16": 8
    }
    
    # Expected for BM.GPU.H100.8: 4x Speed 16GT/s, 8x Speed 32GT/s
    expected_speed_counts = {
        "Speed 16GT/s": 4,
        "Speed 32GT/s": 8
    }
    
    error_messages = []
    
    # Check width counts
    for width, expected_count in expected_width_counts.items():
        actual_count = width_counts.get(width, 0)
        if actual_count != expected_count:
            error_messages.append(f"GPU/NVSwitch PCIe width mismatch: expected {expected_count}x {width}, got {actual_count}x")
    
    # Check speed counts
    for speed, expected_count in expected_speed_counts.items():
        actual_count = speed_counts.get(speed, 0)
        if actual_count != expected_count:
            error_messages.append(f"GPU/NVSwitch PCIe speed mismatch: expected {expected_count}x {speed}, got {actual_count}x")
    
    # Add state errors
    if state_errors:
        error_messages.extend([f"GPU/NVSwitch state error: {err}" for err in state_errors])
    
    if error_messages:
        return False, "; ".join(error_messages), width_counts, speed_counts, state_errors
    
    return True, "", width_counts, speed_counts, state_errors


def check_rdma_pcie_width() -> Tuple[bool, str, Dict[str, int], Dict[str, int], List[str]]:
    """
    Check PCIe width, speed, and state for RDMA interfaces.
    
    Returns:
        Tuple of (success, error_message, width_counts, speed_counts, state_errors)
    """
    cmd = "sudo lspci -vvv | egrep '^[0-9,a-f]|LnkSta:' | grep -A 1 -i mellanox | grep LnkSta | sort | uniq -c"
    
    exit_code, stdout, stderr = run_command(cmd)
    
    if exit_code != 0:
        return False, f"Failed to execute lspci command: {stderr}", {}, {}, []
    
    if not stdout.strip():
        return False, "No Mellanox PCIe devices found", {}, {}, []
    
    width_counts, speed_counts, state_errors = parse_lspci_width_output(stdout)
    
    # Expected for BM.GPU.H100.8: 2x Width x8, 16x Width x16
    expected_width_counts = {
        "Width x8": 2,
        "Width x16": 16
    }
    
    # Expected for BM.GPU.H100.8: 2x Speed 16GT/s, 16x Speed 32GT/s
    expected_speed_counts = {
        "Speed 16GT/s": 2,
        "Speed 32GT/s": 16
    }
    
    error_messages = []
    
    # Check width counts
    for width, expected_count in expected_width_counts.items():
        actual_count = width_counts.get(width, 0)
        if actual_count != expected_count:
            error_messages.append(f"RDMA PCIe width mismatch: expected {expected_count}x {width}, got {actual_count}x")
    
    # Check speed counts
    for speed, expected_count in expected_speed_counts.items():
        actual_count = speed_counts.get(speed, 0)
        if actual_count != expected_count:
            error_messages.append(f"RDMA PCIe speed mismatch: expected {expected_count}x {speed}, got {actual_count}x")
    
    # Add state errors
    if state_errors:
        error_messages.extend([f"RDMA state error: {err}" for err in state_errors])
    
    if error_messages:
        return False, "; ".join(error_messages), width_counts, speed_counts, state_errors
    
    return True, "", width_counts, speed_counts, state_errors


def get_oci_shape() -> str:
    """
    Get the current OCI shape from IMDS or environment variable.
    
    Returns:
        OCI shape string
    """
    # First try environment variable (useful for testing)
    shape = os.environ.get("OCI_SHAPE")
    if shape:
        return shape
    
    # Try IMDS
    try:
        cmd = "curl -s -m 10 http://169.254.169.254/opc/v1/instance/shape"
        exit_code, stdout, stderr = run_command(cmd)
        if exit_code == 0 and stdout.strip():
            return stdout.strip()
    except Exception:
        pass
    
    return "UNKNOWN"


def main():
    """Main function to run PCIe width missing lanes check."""
    
    overall_success = True
    error_messages = []
    
    # Check GPU/NVSwitch PCIe width, speed, and state
    gpu_success, gpu_error, gpu_width_counts, gpu_speed_counts, gpu_state_errors = check_gpu_nvswitch_pcie_width()
    
    if not gpu_success:
        overall_success = False
        error_messages.append(f"GPU/NVSwitch: {gpu_error}")
    
    # Check RDMA PCIe width, speed, and state
    rdma_success, rdma_error, rdma_width_counts, rdma_speed_counts, rdma_state_errors = check_rdma_pcie_width()
    
    if not rdma_success:
        overall_success = False
        error_messages.append(f"RDMA: {rdma_error}")
    
    # Create result in simple format matching other scripts
    if overall_success:
        status = "PASS"
    else:
        status = f"FAIL - {'; '.join(error_messages)}"
    
    result = {
        "pcie_width_missing_lanes": {
            "status": status
        }
    }
    
    print(json.dumps(result, indent=2))
    
    # Exit with appropriate code
    return 0 if overall_success else 1


if __name__ == "__main__":
    sys.exit(main())